File: ontrack-job/src/main/java/net/nemerosa/ontrack/job/support/DefaultJobScheduler.java
package net.nemerosa.ontrack.job.support;

import net.nemerosa.ontrack.common.Time;
import net.nemerosa.ontrack.job.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.Collection;
import java.util.Map;
import java.util.TreeMap;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;
import java.util.stream.Collectors;

public class DefaultJobScheduler implements JobScheduler {

    private final Logger logger = LoggerFactory.getLogger(JobScheduler.class);

    private final JobDecorator jobDecorator;
    private final ScheduledExecutorService scheduledExecutorService;

    private final Map<JobKey, JobScheduledService> services = new ConcurrentHashMap<>(new TreeMap<>());
    private final AtomicBoolean schedulerPaused = new AtomicBoolean(false);

    public DefaultJobScheduler(JobDecorator jobDecorator, ScheduledExecutorService scheduledExecutorService) {
        this.jobDecorator = jobDecorator;
        this.scheduledExecutorService = scheduledExecutorService;
    }

    @Override
    public void schedule(Job job, Schedule schedule) {
        logger.info("[job][{}][{}] Scheduling with {}", job.getKey().getType(), job.getKey().getId(), schedule);
        // Manages existing schedule
        JobScheduledService existingService = services.remove(job.getKey());
        if (existingService != null) {
            logger.info("[job][{}][{}] Stopping existing schedule", job.getKey().getType(), job.getKey().getId());
            existingService.cancel(false);
        }
        // Gets the job task
        Runnable jobTask = job.getTask();
        // Decorates the task
        Runnable decoratedTask = jobDecorator.decorate(job, jobTask);
        // Creates and starts the scheduled service
        logger.info("[job][{}][{}] Starting service", job.getKey().getType(), job.getKey().getId());
        // Copy stats from old schedule
        JobScheduledService jobScheduledService = new JobScheduledService(job, decoratedTask, schedule, scheduledExecutorService, existingService);
        // Registration
        services.put(job.getKey(), jobScheduledService);
    }

    @Override
    public void unschedule(JobKey key) {
        JobScheduledService existingService = services.remove(key);
        if (existingService != null) {
            existingService.cancel(true);
        }
    }

    @Override
    public void pause() {
        schedulerPaused.set(true);
    }

    @Override
    public void resume() {
        schedulerPaused.set(false);
    }

    @Override
    public void pause(JobKey key) {
        JobScheduledService existingService = services.get(key);
        if (existingService != null) {
            existingService.pause();
        } else {
            throw new JobNotScheduledException(key);
        }
    }

    @Override
    public void resume(JobKey key) {
        JobScheduledService existingService = services.get(key);
        if (existingService != null) {
            existingService.resume();
        } else {
            throw new JobNotScheduledException(key);
        }
    }

    @Override
    public JobStatus getJobStatus(JobKey key) {
        JobScheduledService existingService = services.get(key);
        if (existingService != null) {
            return existingService.getJobStatus();
        } else {
            throw new JobNotScheduledException(key);
        }
    }

    @Override
    public Collection<JobKey> getAllJobKeys() {
        return services.keySet();
    }

    @Override
    public Collection<JobKey> getJobKeysOfType(String type) {
        return getAllJobKeys().stream()
                .filter(key -> key.sameType(type))
                .collect(Collectors.toSet());
    }

    @Override
    public Collection<JobStatus> getJobStatuses() {
        return services.values().stream()
                .map(JobScheduledService::getJobStatus)
                .collect(Collectors.toList());
    }

    @Override
    public Future<?> fireImmediately(JobKey jobKey) {
        // Gets the existing scheduled service
        JobScheduledService jobScheduledService = services.get(jobKey);
        if (jobScheduledService == null) {
            throw new JobNotScheduledException(jobKey);
        }
        // Fires the job immediately
        return jobScheduledService.fireImmediately();
    }

    private class JobScheduledService implements Runnable {

        private final Job job;
        private final Schedule schedule;
        private final Runnable monitoredTask;
        private final ScheduledFuture<?> scheduledFuture;

        private final AtomicBoolean paused = new AtomicBoolean(false);
        private final AtomicReference<CompletableFuture<?>> completableFuture = new AtomicReference<>();

        private final AtomicLong runCount = new AtomicLong();
        private final AtomicReference<LocalDateTime> lastRunDate = new AtomicReference<>();
        private final AtomicLong lastRunDurationMs = new AtomicLong();
        private final AtomicLong lastErrorCount = new AtomicLong();
        private final AtomicReference<String> lastError = new AtomicReference<>(null);

        private JobScheduledService(Job job, Runnable decoratedTask, Schedule schedule, ScheduledExecutorService scheduledExecutorService, JobScheduledService old) {
            this.job = job;
            this.schedule = schedule;
            this.monitoredTask = new MonitoredTask(decoratedTask);
            // Copies stats from old service
            if (old != null) {
                runCount.set(old.runCount.get());
                lastRunDate.set(old.lastRunDate.get());
                lastRunDurationMs.set(old.lastRunDurationMs.get());
                lastErrorCount.set(old.lastErrorCount.get());
                lastError.set(old.lastError.get());
            }
            // Scheduling now
            scheduledFuture = scheduledExecutorService.scheduleWithFixedDelay(
                    this,
                    schedule.getInitialPeriod(),
                    schedule.getPeriod(),
                    schedule.getUnit()
            );
        }

        @Override
        public void run() {
            fireImmediately();
        }

        public void cancel(boolean forceStop) {
            scheduledFuture.cancel(false);
            // The decorated task might still run
            CompletableFuture<?> future = this.completableFuture.get();
            if (forceStop && future != null && !future.isDone() && !future.isCancelled()) {
                future.cancel(true);
            }
        }

        public Future<?> fireImmediately() {
            return completableFuture.updateAndGet(this::optionallyFireTask);
        }

        protected CompletableFuture<?> optionallyFireTask(CompletableFuture<?> runningCompletableFuture) {
            if (runningCompletableFuture != null) {
                return runningCompletableFuture;
            } else {
                return fireTask();
            }
        }

        protected CompletableFuture<Void> fireTask() {
            return CompletableFuture
                    .runAsync(monitoredTask, scheduledExecutorService)
                    .whenComplete((ignored, ex) -> completableFuture.set(null));
        }

        public JobStatus getJobStatus() {
            return new JobStatus(
                    job.getKey(),
                    schedule,
                    job.getDescription(),
                    completableFuture.get() != null,
                    runCount.get(),
                    lastRunDate.get(),
                    lastRunDurationMs.get(),
                    getNextRunDate(),
                    lastErrorCount.get(),
                    lastError.get()
            );
        }

        private LocalDateTime getNextRunDate() {
            LocalDateTime date = lastRunDate.get();
            if (date != null) {
                return date.plus(schedule.toMiliseconds(), ChronoUnit.MILLIS);
            } else {
                return null;
            }
        }

        public void pause() {
            paused.set(true);
        }

        public void resume() {
            paused.set(false);
        }

        private class MonitoredTask implements Runnable {

            private final Runnable decoratedTask;

            public MonitoredTask(Runnable decoratedTask) {
                this.decoratedTask = decoratedTask;
            }

            @Override
            public void run() {
                if (isEnabled()) {
                    try {
                        logger.debug("[job][{}][{}] Running now", job.getKey().getType(), job.getKey().getId());
                        lastRunDate.set(Time.now());
                        runCount.incrementAndGet();
                        // Runs the job
                        long _start = System.currentTimeMillis();
                        decoratedTask.run();
                        // No error, counting time
                        long _end = System.currentTimeMillis();
                        lastRunDurationMs.set(_end - _start);
                        logger.debug("[job][{}][{}] Ran in {} ms", job.getKey().getType(), job.getKey().getId(), lastRunDurationMs.get());
                        // No error - resetting the counters
                        lastErrorCount.set(0);
                        lastError.set(null);
                    } catch (RuntimeException ex) {
                        lastErrorCount.incrementAndGet();
                        lastError.set(ex.getMessage());
                        logger.error("[job][{}][{}] Error: {}", job.getKey().getType(), job.getKey().getId(), ex.getMessage());
                        // TODO Error reporter
                        throw ex;
                    }
                } else {
                    logger.debug("[job][{}][{}] Not enabled", job.getKey().getType(), job.getKey().getId());
                }
            }
        }

        private boolean isEnabled() {
            return !job.isDisabled() && !paused.get() && !schedulerPaused.get();
        }
    }
}

File: ontrack-job/src/main/java/net/nemerosa/ontrack/job/support/DefaultJobScheduler.kt
package net.nemerosa.ontrack.job.support

import net.nemerosa.ontrack.common.Time
import net.nemerosa.ontrack.job.*
import org.apache.commons.lang3.Validate
import org.slf4j.LoggerFactory
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit
import java.util.*
import java.util.concurrent.*
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.atomic.AtomicReference
import java.util.function.BiFunction

class DefaultJobScheduler(
        private val jobDecorator: JobDecorator,
        private val schedulerPool: ScheduledExecutorService,
        private val jobListener: JobListener,
        initiallyPaused: Boolean,
        private val jobPoolProvider: BiFunction<ExecutorService, Job, ExecutorService>,
        private val scattering: Boolean,
        private val scatteringRatio: Double
) : JobScheduler {

    private val logger = LoggerFactory.getLogger(JobScheduler::class.java)

    private val services = ConcurrentHashMap(TreeMap<JobKey, JobScheduledService>())
    private val schedulerPaused: AtomicBoolean

    private val idGenerator = AtomicLong()

    constructor(
            jobDecorator: JobDecorator,
            schedulerPool: ScheduledExecutorService,
            jobListener: JobListener,
            initiallyPaused: Boolean,
            scattering: Boolean,
            scatteringRatio: Double
    ) : this(
            jobDecorator,
            schedulerPool,
            jobListener,
            initiallyPaused,
            BiFunction { executorService, _ -> executorService },
            scattering,
            scatteringRatio
    )

    init {
        Validate.inclusiveBetween(0.0, 1.0, scatteringRatio)
        this.schedulerPaused = AtomicBoolean(initiallyPaused)
    }

    override fun schedule(job: Job, schedule: Schedule) {
        logger.info("[scheduler][job]{} Scheduling with {}", job.key, schedule)
        // Manages existing schedule
        val existingService = services.remove(job.key)
        if (existingService != null) {
            logger.info("[scheduler][job]{} Stopping existing schedule", job.key)
            existingService.cancel(false)
        }
        // Creates and starts the scheduled service
        logger.info("[scheduler][job]{} Starting scheduled service", job.key)
        // Copy stats from old schedule
        val jobScheduledService = JobScheduledService(
                job,
                schedule,
                schedulerPool,
                existingService,
                jobListener.isPausedAtStartup(job.key)
        )
        // Registration
        services[job.key] = jobScheduledService
    }

    override fun unschedule(key: JobKey): Boolean {
        return unschedule(key, true)
    }

    private fun unschedule(key: JobKey, forceStop: Boolean): Boolean {
        logger.debug("[scheduler][job]{} Unscheduling job", key)
        val existingService = services.remove(key)
        return if (existingService != null) {
            logger.debug("[scheduler][job]{} Stopping running job", key)
            existingService.cancel(forceStop)
            true
        } else {
            false
        }
    }

    override fun pause() {
        schedulerPaused.set(true)
    }

    override fun resume() {
        schedulerPaused.set(false)
    }

    override fun isPaused(): Boolean {
        return schedulerPaused.get()
    }

    override fun pause(key: JobKey): Boolean {
        val existingService = services[key]
        if (existingService != null) {
            existingService.pause()
            return true
        } else {
            throw JobNotScheduledException(key)
        }
    }

    override fun resume(key: JobKey): Boolean {
        val existingService = services[key]
        if (existingService != null) {
            existingService.resume()
            return true
        } else {
            throw JobNotScheduledException(key)
        }
    }

    override fun getJobStatus(key: JobKey): Optional<JobStatus> {
        val existingService = services[key]
        return if (existingService != null) {
            Optional.of(existingService.jobStatus)
        } else {
            Optional.empty()
        }
    }

    override fun getJobKey(id: Long): Optional<JobKey> {
        return services.values.stream()
                .filter { service -> service.id == id }
                .map { it.jobKey }
                .findFirst()
    }

    override fun stop(key: JobKey): Boolean {
        val existingService = services[key]
        return existingService?.stop() ?: throw JobNotScheduledException(key)
    }

    override fun getAllJobKeys(): Collection<JobKey> {
        return services.keys
    }

    override fun getJobKeysOfType(type: JobType): Collection<JobKey> {
        return allJobKeys
                .filter { key -> key.sameType(type) }
                .toSet()
    }

    override fun getJobKeysOfCategory(category: JobCategory): Collection<JobKey> {
        return allJobKeys
                .filter { key -> key.sameCategory(category) }
                .toSet()
    }

    override fun getJobStatuses(): Collection<JobStatus> {
        return services.values
                .map { it.jobStatus }
                .sortedBy { it.id }
                .toList()
    }

    override fun fireImmediately(jobKey: JobKey): Optional<Future<*>> {
        // Gets the existing scheduled service
        val jobScheduledService = services[jobKey] ?: throw JobNotScheduledException(jobKey)
        // Fires the job immediately
        return jobScheduledService.doRun(true)
    }

    private fun getExecutorService(job: Job): ExecutorService {
        return jobPoolProvider.apply(schedulerPool, job)
    }

    private inner class JobScheduledService(
            private val job: Job,
            private val schedule: Schedule,
            scheduledExecutorService: ScheduledExecutorService,
            old: JobScheduledService?,
            pausedAtStartup: Boolean
    ) : Runnable {

        val id: Long = idGenerator.incrementAndGet()
        private val actualSchedule: Schedule
        private val scheduledFuture: ScheduledFuture<*>?

        private val paused: AtomicBoolean = AtomicBoolean(pausedAtStartup)

        private val currentExecution = AtomicReference<Future<*>>()
        private val runProgress = AtomicReference<JobRunProgress>()
        private val runCount = AtomicLong()
        private val lastRunDate = AtomicReference<LocalDateTime>()
        private val lastRunDurationMs = AtomicLong()
        private val lastErrorCount = AtomicLong()
        private val lastError = AtomicReference<String>(null)

        val jobKey: JobKey
            get() = job.key

        private val run: Runnable
            get() {
                val jobRunListener = DefaultJobRunListener()
                val rootTask = { job.task.run(jobRunListener) }
                val decoratedTask = jobDecorator.decorate(job, rootTask)
                val runnable = MonitoredRun(decoratedTask, object : MonitoredRunListenerAdapter() {
                    override fun onCompletion() {
                        logger.debug("[job][task]{} Removed job execution", job.key)
                        currentExecution.set(null)
                    }
                })
                val monitoredRunListener = object : MonitoredRunListener {
                    override fun onStart() {
                        logger.debug("[job][task]{} On start", job.key)
                        lastRunDate.set(Time.now())
                        runCount.incrementAndGet()
                        jobListener.onJobStart(job.key)
                    }

                    override fun onSuccess(duration: Long) {
                        lastRunDurationMs.set(duration)
                        logger.debug("[job][task]{} Success in {} ms", job.key, duration)
                        jobListener.onJobEnd(job.key, duration)
                        lastErrorCount.set(0)
                        lastError.set(null)
                    }

                    override fun onFailure(ex: Exception) {
                        lastErrorCount.incrementAndGet()
                        lastError.set(ex.message)
                        logger.debug("[job][task]{} Failure: {}", job.key, ex.message)
                        jobListener.onJobError(jobStatus, ex)
                    }

                    override fun onCompletion() {
                        runProgress.set(null)
                        logger.debug("[job][task]{} Job completed.", job.key)
                        jobListener.onJobComplete(job.key)
                    }
                }
                return MonitoredRun(runnable, monitoredRunListener)
            }

        val jobStatus: JobStatus
            get() {
                val valid = job.isValid
                return JobStatus(
                        id,
                        job.key,
                        schedule,
                        actualSchedule,
                        job.description,
                        currentExecution.get() != null,
                        valid,
                        paused.get(),
                        job.isDisabled,
                        runProgress.get(),
                        runCount.get(),
                        lastRunDate.get(),
                        lastRunDurationMs.get(),
                        getNextRunDate(valid),
                        lastErrorCount.get(),
                        lastError.get()
                )
            }

        init {
            // Paused at startup
            if (pausedAtStartup) {
                logger.debug("[job]{} Job paused at startup", job.key)
            }
            // Copies stats from old service
            if (old != null) {
                runCount.set(old.runCount.get())
                lastRunDate.set(old.lastRunDate.get())
                lastRunDurationMs.set(old.lastRunDurationMs.get())
                lastErrorCount.set(old.lastErrorCount.get())
                lastError.set(old.lastError.get())
            }
            // Converting all units to milliseconds
            var initialPeriod = TimeUnit.MILLISECONDS.convert(schedule.initialPeriod, schedule.unit)
            val period = TimeUnit.MILLISECONDS.convert(schedule.period, schedule.unit)
            // Scattering
            if (scattering) {
                // Computes the hash for the job key
                val hash = Math.abs(job.key.toString().hashCode()) % 10000
                // Period to consider
                val scatteringMax = (period * scatteringRatio).toLong()
                if (scatteringMax > 0) {
                    // Modulo on the period
                    val delay = hash * scatteringMax / 10000
                    logger.debug("[job]{} Scattering enabled - additional delay: {} ms", job.key, delay)
                    // Adding to the initial delay
                    initialPeriod += delay
                }
            }
            // Actual schedule
            actualSchedule = Schedule(
                    initialPeriod,
                    period,
                    TimeUnit.MILLISECONDS
            )
            // Scheduling now
            scheduledFuture = if (schedule.period > 0) {
                scheduledExecutorService.scheduleWithFixedDelay(
                        this,
                        initialPeriod,
                        period,
                        TimeUnit.MILLISECONDS
                )
            } else {
                logger.debug("[job]{} Job not scheduled since period = 0", job.key)
                null
            }
        }

        override fun run() {
            if (!schedulerPaused.get()) {
                doRun(false)
            }
        }

        fun doRun(force: Boolean): Optional<Future<*>> {
            logger.debug("[job][run]{} Trying to run now - forced = {}", job.key, force)
            if (job.isValid) {
                if (job.isDisabled) {
                    logger.debug("[job][run]{} Not allowed to run now because disabled", job.key)
                    return Optional.empty()
                } else if (paused.get() && !force) {
                    logger.debug("[job][run]{} Not allowed to run now because paused", job.key)
                    return Optional.empty()
                } else if (currentExecution.get() != null) {
                    logger.debug("[job][run]{} Not allowed to run now because already running", job.key)
                    return Optional.empty()
                } else {
                    // Task to run
                    val run = run
                    // Gets the executor for this job
                    val executor = getExecutorService(job)
                    // Scheduling
                    logger.debug("[job][run]{} Job task submitted asynchronously", job.key)
                    val execution = executor.submit(run)
                    currentExecution.set(execution)
                    return Optional.of(execution)
                }
            } else {
                logger.debug("[job][run]{} Not valid - removing from schedule", job.key)
                unschedule(job.key, false)
                return Optional.empty()
            }
        }

        fun stop(): Boolean {
            logger.debug("[job]{} Stopping job", job.key)
            return currentExecution.updateAndGet { current ->
                current?.cancel(true)
                null
            } == null
        }

        fun cancel(forceStop: Boolean): Boolean {
            logger.debug("[job]{} Cancelling job (forcing = {})", job.key, forceStop)
            if (forceStop) {
                stop()
            }
            return scheduledFuture != null && scheduledFuture.cancel(forceStop)
        }

        private fun getNextRunDate(valid: Boolean): LocalDateTime? {
            return if (valid && scheduledFuture != null) {
                Time.now().plus(
                        scheduledFuture.getDelay(TimeUnit.SECONDS),
                        ChronoUnit.SECONDS
                )
            } else {
                null
            }
        }

        fun pause() {
            if (scheduledFuture != null) {
                paused.set(true)
                jobListener.onJobPaused(job.key)
            }
        }

        fun resume() {
            if (scheduledFuture != null) {
                paused.set(false)
                jobListener.onJobResumed(job.key)
            }
        }

        private inner class DefaultJobRunListener : JobRunListener {

            override fun progress(progress: JobRunProgress) {
                jobListener.onJobProgress(job.key, progress)
                logger.debug("[job][progress]{} {}",
                        job.key,
                        progress.text
                )
                runProgress.set(progress)
            }

        }

    }
}
